### PR TITLE
Rework the action runner

### LINKED ISSUE #46: The logic of searching for consecutive replacements is broken
The current logic here:

https://github.com/platisd/clang-tidy-pr-comments/blob/3fae50b904c5c9fd507b9fab1cb916055a392e8c/run_action.py#L218-L233

is problematic. Consider the following `fixes.yml`:

```yaml
Diagnostics:
- BuildDirectory: /home/runner/work/fheroes2/fheroes2/build/src/fheroes2
  DiagnosticMessage:
    FileOffset: 7837
    FilePath: /home/runner/work/fheroes2/fheroes2/src/fheroes2/kingdom/kingdom.cpp
    Message: return type 'const EventsDate' (aka 'const list<EventDate>') is 'const'-qualified
      at the top level, which may reduce code readability without improving const
      correctness
    Ranges:
    - FileOffset: 7837
      FilePath: /home/runner/work/fheroes2/fheroes2/src/fheroes2/kingdom/kingdom.cpp
      Length: 5
    Replacements:
    - FilePath: /home/runner/work/fheroes2/fheroes2/src/fheroes2/kingdom/kingdom.cpp
      Length: 6
      Offset: 7837
      ReplacementText: ''
    - FilePath: /home/runner/work/fheroes2/fheroes2/src/fheroes2/kingdom/kingdom.h
      Length: 6
      Offset: 4838
      ReplacementText: ''
  DiagnosticName: readability-const-return-type
  Level: Warning
MainSourceFile: ''
```

First of all, these replacements are for different files with different `FilePath`s in each replacement, but this is not taken into account when checking for continuity. Second, even though these replacements are not consecutive, they are still considered as such, because `7837 + 6` is not less than `4838 - 1`.

### LINKED ISSUE #47: Failed Assertion `run_action.py:244`
My repo is failing this assertion

```python
assert all(path == file_paths[0] for path in file_paths)
```

<details>
  <summary>My Workflow yml</summary>
  
  ```yml
  name: idf

  on:
    pull_request:
    push:
      branches:
        - main
  
  jobs:
    clang-tidy:
      runs-on: ubuntu-latest
      container:
        image: espressif/idf:release-v5.1
      steps:
        - uses: actions/checkout@v2
          with:
            submodules: true
  
        - name: Install tools
          run: |
            . $IDF_PATH/export.sh
            idf_tools.py install esp-clang
            pip install --upgrade pyclang
  
        - name: Run clang-tidy
          run: |
            . $IDF_PATH/export.sh
            idf.py clang-check --run-clang-tidy-options "-export-fixes build/fixes.yml -header-filter=main/.*" --exclude-paths managed_components
  
        - name: Run clang-tidy-pr-comments action
          uses: platisd/clang-tidy-pr-comments@master
          with:
            github_token: ${{ secrets.GITHUB_TOKEN }}
            clang_tidy_fixes: build/fixes.yml
            request_changes: true
            suggestions_per_comment: 10
            repo_path_prefix: "/__w"
  
        - uses: actions/upload-artifact@v3
          if: always()
          with:
            name: fixes
            path: build/fixes.yml

  ```

</details>

<details>
  <summary>My build/fixes.yml (`head -n 50` -- full version attached)</summary>
  
  ```yml
  Diagnostics:
  - BuildDirectory: /__w/mainframe-tester/mainframe-tester/build
    DiagnosticMessage:
      FileOffset: 95
      FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.hpp
      Message: invalid case style for global constant 'kDut3_3V'
      Replacements:
      - FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.hpp
        Length: 8
        Offset: 95
        ReplacementText: kDut33V
      - FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.cpp
        Length: 8
        Offset: 82
        ReplacementText: kDut33V
    DiagnosticName: readability-identifier-naming
    Level: Warning
  - BuildDirectory: /__w/mainframe-tester/mainframe-tester/build
    DiagnosticMessage:
      FileOffset: 1481
      FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.hpp
      Message: invalid case style for global function 'initGpio'
      Replacements:
      - FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.hpp
        Length: 8
        Offset: 1481
        ReplacementText: InitGpio
      - FilePath: /__w/mainframe-tester/mainframe-tester/main/GPIO.cpp
        Length: 8
        Offset: 46
        ReplacementText: InitGpio
    DiagnosticName: readability-identifier-naming
    Level: Warning
  - BuildDirectory: /__w/mainframe-tester/mainframe-tester/build
    DiagnosticMessage:
      FileOffset: 35
      FilePath: /__w/mainframe-tester/mainframe-tester/main/NVS.hpp
      Message: invalid case style for global function 'initNvs'
      Replacements:
      - FilePath: /__w/mainframe-tester/mainframe-tester/main/NVS.hpp
        Length: 7
        Offset: 35
        ReplacementText: InitNvs
    DiagnosticName: readability-identifier-naming
    Level: Warning
  - BuildDirectory: /__w/mainframe-tester/mainframe-tester/build
    DiagnosticMessage:
      FileOffset: 52
      FilePath: /__w/mainframe-tester/mainframe-tester/main/NVS.hpp
      Message: invalid case style for global function 'swap_servo_direction'
  ```

</details>

[fixes.yml.zip](https://github.com/platisd/clang-tidy-pr-comments/files/11912702/fixes.yml.zip)

Nothing seems out of place in my paths. The `__w` part is because I'm indeed running clang-tidy inside docker, but seems like that is what the `repo_path_prefix` is meant to handle.

I added a simple bit of printing to [my fork](https://github.com/johnboiles/clang-tidy-pr-comments/blob/master/run_action.py#L244-L245) before the assert:

```python
for path in file_paths:                
    print(f"{path} == {file_paths[0]}")
```

And I see this in the logs:

```
main/mainframe-tester.cpp == main/mainframe-tester.cpp
main/ADC.hpp == main/mainframe-tester.cpp
```

So it would seem that the assertion failure is coming from this replacement:

```
- BuildDirectory: /__w/mainframe-tester/mainframe-tester/build
  DiagnosticMessage:
    FileOffset: 5
    FilePath: /__w/mainframe-tester/mainframe-tester/main/ADC.hpp
    Message: invalid case style for global function 'initAdc'
    Replacements:
    - FilePath: /__w/mainframe-tester/mainframe-tester/main/mainframe-tester.cpp
      Length: 7
      Offset: 4544
      ReplacementText: InitAdc
    - FilePath: /__w/mainframe-tester/mainframe-tester/main/ADC.hpp
      Length: 7
      Offset: 5
      ReplacementText: InitAdc
  DiagnosticName: readability-identifier-naming
  Level: Warning
```

This replacement seems valid, since it's changing a method name in both the header where it's defined and in a `.cpp` file where it's used.

It appears the code in its current state assumes that each replacement is only in a single file, which seems to not necessarily be the case, at least in clang-tidy 15.

### PR DESCRIPTION
**Pull request by oleg-derevenetz**
fix #46
fix #47

This PR offers an almost completely rewritten runner script based on new principles. Here is an example of its work:

https://github.com/oleg-derevenetz/fheroes2/pull/160
https://github.com/oleg-derevenetz/fheroes2/actions/runs/6750999770/job/18354374497#step:8:15

Highlights:

* Completely new logic for working with suggestions. Multiple changes in one line, in adjacent lines or in different files are no longer a problem.
* Improved logging.
* Modernized code (with the assistance of `pylint`).

### CODE REVIEW

**Comment by platisd** [comment]
Wow you're a hero! :rocket: :heart: 
Taking a look.

File: run_action.py [hunk 1]
```diff
@@ -13,492 +16,593 @@
...
+                                )
+
+                            start_line_num = None
+                            end_line_num = None
+                            replacement_text = None
+
+                        line_num += 1
+                    # Unknown prefix, this should not happen
+                    else:
+                        assert False
```
Comment (platisd) [inline run_action.py:348]:
```suggestion
                        assert False, "Please report this to the repository maintainer"
```

**Review by platisd — CHANGES_REQUESTED** [review]
I went through it and it looks very good. Your test run also looks like there are no obvious regressions. :100: 
Should probably have something like this as test cases. :tired_face: 

However, it's a bit hard to follow what has stayed the same, since things have moved around.

For clarity and to make sure we haven't missed anything, can you please point out the function(s) that have the bulk of changes in the parsing logic?
Is it primarily `generate_review_comments`?

**Comment by oleg-derevenetz** [comment]
> However, it's a bit hard to follow what has stayed the same, since things have moved around.

Not much really :)

> For clarity and to make sure we haven't missed anything, can you please point out the function(s) that have the bulk of changes in the parsing logic?
Is it primarily `generate_review_comments`?

In fact, yes. But the logic of processing replacements has been radically changed. Now, for each diagnostics with suggestions, all the suggestions from that diagnostics belonging to a particular file are applied to that file and then `difflib` is used to compute the line-by-line diff (similar to unified diff), which is then processed to generate the suggestion diffs suitable for GitHub. This allows generation of just one suggestion for cases of multiple changes in one line or in adjacent lines, and in general, it makes logic easier and simpler.

**Review by platisd — APPROVED** [review]

**Comment by platisd** [comment]
Great job! :handshake: :rocket: :heart: 
I'll bump the release after no one complains for a week or so :grin:
